File: My_Notes/Programming_Note/Samples/async_samples/go/async2.go
package main

import "sync"

func asyncFunc2() {
	job := job{
		db:  3,
		api: 1,
		ui:  2,
	}
	var dbEngineer dbEngineer
	var apiEngineer apiEngineer
	var uiEngineer uiEngineer

	wgAll := sync.WaitGroup{}
	wgAll.Add(3)

	dbDone := make(chan bool)

	go func() {
		dbEngineer.do(job)
		wgAll.Done()
		dbDone <- true
	}()

	go func() {
		if <-dbDone {
			apiEngineer.do(job)
			wgAll.Done()
		}
	}()

	go func() {
		uiEngineer.do(job)
		wgAll.Done()
	}()

	wgAll.Wait()
}

File: My_Notes/Programming_Note/Samples/async_samples/ts/async2.ts
import { DBEngineer, APIEngineer, UIEngineer } from "./engineer";

function async2():Promise<[void,void]>{
    var job = new Job(3,1,2);
    var dbEngineer = new DBEngineer();
    var apiEngineer = new APIEngineer();
    var uiEngineer = new UIEngineer();

    var task1:Promise<void> = dbEngineer.do(job).then(() =>apiEngineer.do(job))
    var task2:Promise<void> = uiEngineer.do(job)

    return Promise.all([task1, task2])
}